org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/Annotation.java

```diff
@@ -1313,7 +1299,7 @@ static AnnotationTargetAllowed isAnnotationTargetAllowed(Annotation annotation,
 		}
 
 		// https://bugs.eclipse.org/bugs/show_bug.cgi?id=391201
-		if ((metaTagBits & TagBits.SE7AnnotationTargetMASK) == 0
+		if ((metaTagBits & TagBits.AnnotationForDeclarationMASK) == 0
 				&& (metaTagBits & (TagBits.AnnotationForTypeUse | TagBits.AnnotationForTypeParameter)) != 0) {
 			if (scope.compilerOptions().sourceLevel < ClassFileConstants.JDK1_8) {
```

Comment:
Should this be:
```
if ((metaTagBits & TagBits.AnnotationForDeclarationMASK) == 0
				&& (metaTagBits & TagBits.AnnotationForTypeUse) != 0) {
```

Reply:
Good catch, I only modified one line without looking into the next ... 

Still I don't think it's possible to construct a case where it would make a difference, as we'd need to provide "no declaration target but target type parameter", which however is a complicated way of saying "false" :)

I'll push a fix, just to avoid future puzzlement.